play-route/src/main/scala/org/http4s/server/play/PlayRouteBuilder.scala

```diff
@@ -0,0 +1,149 @@
+package org.http4s.server.play
+
+import akka.stream.scaladsl.{Sink, Source}
+import akka.util.ByteString
+import cats.data.OptionT
+import cats.effect.{Effect, IO}
+import fs2.interop.reactivestreams._
+import org.http4s.server.play.PlayRouteBuilder.{PlayAccumulator, PlayRouting, PlayTargetStream}
+import org.http4s.{EmptyBody, Header, Headers, HttpService, Method, Request, Response, Uri}
+import play.api.http.HttpEntity.Streamed
+import play.api.libs.streams.Accumulator
+import play.api.mvc._
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, Promise}
+
+class PlayRouteBuilder[F[_]](
+    service: HttpService[F]
+)(implicit F: Effect[F], executionContext: ExecutionContext) {
+
+  type UnwrappedKleisli = Request[F] => OptionT[F, Response[F]]
+  private[this] val unwrappedRun: UnwrappedKleisli = service.run
+
+  def requestHeaderToRequest(requestHeader: RequestHeader): Request[F] =
+    Request(
+      // todo fix the ???
+      method = Method.fromString(requestHeader.method).getOrElse(???),
+      uri = Uri(path = requestHeader.uri),
+      headers = Headers.apply(requestHeader.headers.toMap.toList.flatMap {
+        case (headerName, values) =>
+          values.map { value =>
+            Header(headerName, value)
+          }
+      }),
+      body = EmptyBody
+    )
+
+  type ResponseStream = fs2.Stream[F, Byte]
+
+  def convertStream(responseStream: ResponseStream): PlayTargetStream = {
+    val entityBody: fs2.Stream[F, Byte] = responseStream
+    Source
+      .fromPublisher(entityBody.toUnicastPublisher())
+      .map(byte => ByteString(byte))
+  }
+
+  /**
+    * A Play accumulator Sinks HTTP data in, and then pumps out a future of a Result.
+    * That Result will have a Source as the response HTTP Entity.
+    *
+    * Here we create a unattached sink, map its materialized value into a publisher,
+    * convert that into an FS2 Stream, then pipe the request body into the http4s request.
+    */
+  def playRequestToPlayResponse(requestHeader: RequestHeader): PlayAccumulator = {
+    val sink: Sink[ByteString, Future[Result]] = {
+      Sink.asPublisher[ByteString](false).mapMaterializedValue { publisher =>
+        val requestBodyStream: fs2.Stream[F, Byte] =
+          publisher.toStream().flatMap(bs => fs2.Stream.fromIterator[F, Byte](bs.toIterator))
+
+        val wrappedResponse: F[Response[F]] =
+          F.map(unwrappedRun(
+            requestHeaderToRequest(requestHeader).withBodyStream(requestBodyStream)).value)(_.get)
```

Comment:
That `(_.get)` is going to throw if there's no matching route.  There's a `Response.notFound` for a canned 404 response.

Also note that in master, we're moving toward an `HttpApp` type, which will eliminate this complication of optionality.  But you're on the right track for 0.18.x.